Tests/scripts/add_secrets_file_to_build.py

```diff
@@ -1,27 +1,160 @@
+                continue
+
+        if git_status not in {'A', 'M', 'D', }:
+            logging.error(f'unexpected {git_status=}, considering it as <M>odified')
+
+        changed_files.append(file_path)
 
+    return changed_files
 
-def run(options):
+
+def get_secrets_from_gsm(branch_name: str, options: argparse.Namespace, yml_pack_ids: list[str]) -> dict:
+    """
+    Gets the dev secrets and main secrets from GSM and merges them
+    :param branch_name: the name of the branch of the PR
+    :param options: the parsed parameter for the script
+    :param yml_pack_ids: a list of IDs of changed integrations
+    :return: the list of secrets from GSM to use in the build
+    """
     secret_conf = GoogleSecreteManagerModule(options.service_account)
-    secrets = secret_conf.list_secrets(options.gsm_project_id, with_secret=True, attr_validation=('name', 'params'))
+    labels_filter_branch = {'pack_id': '!=None', 'ignore': '==None', 'merged': '==None', 'dev': '==None'}
+    labels_filter_master = {'pack_id': '!=None', 'ignore': '==None', 'merged': '==None', 'dev': '!=None',
+                            'branch_name': f'=="{branch_name}"'}
+    master_secrets = secret_conf.list_secrets(options.gsm_project_id, labels_filter_branch, name_filter=yml_pack_ids,
+                                              with_secrets=True)
+    branch_secrets = secret_conf.list_secrets(options.gsm_project_id, labels_filter_master, with_secrets=True)
+
+    if branch_secrets:
+        for dev_secret in branch_secrets:
+            replaced = False
+            for i in range(len(master_secrets)):
+                if dev_secret['name'] == master_secrets[i]['name']:
+                    master_secrets[i] = dev_secret
+                    replaced = True
+                    break
+            # If the dev secret is not in the changed packs it's a new secret
+            if not replaced:
+                master_secrets.append(dev_secret)
+
     secret_file = {
         "username": options.user,
         "userPassword": options.password,
-        "integrations": secrets
+        "integrations": master_secrets
     }
+    return secret_file
 
+
+def write_secrets_to_file(options: argparse.Namespace, secrets: dict):
+    """
+    Writes the secrets we got from GSM to a file for the build
+    :param options: the parsed parameter for the script
+    :param secrets: a list of secrets to be used in the build
+    """
     with open(options.json_path_file, 'w') as secrets_out_file:
         try:
-            secrets_out_file.write(json5.dumps(secret_file, quote_keys=True))
+            secrets_out_file.write(json5.dumps(secrets, quote_keys=True))
         except Exception as e:
             logging.error(f'Could not save secrets file, malformed json5 format, the error is: {e}')
     logging.info(f'saved the json file to: {options.json_path_file}')
 
 
-def options_handler(args=None):
+def get_yml_pack_ids(changed_packs: list[str]) -> list[str]:
```

Comment:
We have a bunch of solutions for this function, please consult with @ilaner to see whether we can integrate an existing solution.